pkg/streamingpromql/operators/selectors/extend_range_vector.go

```diff
@@ -0,0 +1,192 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+// Provenance-includes-location: https://github.com/prometheus/prometheus/blob/main/promql/engine.go
+// Provenance-includes-license: Apache-2.0
+// Provenance-includes-copyright: The Prometheus Authors
+
+package selectors
+
+import (
+	"github.com/prometheus/prometheus/promql"
+
+	"github.com/grafana/mimir/pkg/streamingpromql/types"
+	"github.com/grafana/mimir/pkg/util/limiter"
+)
+
+type ExtendedPoints struct {
+	points []promql.FPoint
+}
+
+type ExtendedSmoothedPoints struct {
+	ExtendedPoints
+	smoothedHeadSet bool
+	smoothedTailSet bool
+}
+
+type ExtendedRangeVectorPoints struct {
+	points ExtendedPoints
+	first  promql.FPoint
+	last   promql.FPoint
+}
+
+type ExtendedSmoothedRangeVectorPoints struct {
+	ExtendedRangeVectorPoints
+	smoothedHead *promql.FPoint
+	smoothedTail *promql.FPoint
+}
+
+// NewExtendedPointsForAnchored returns a slice of points adjusted to include
+// anchored boundary points at the start and end of the specified range.
+//
+// A synthetic point is placed at both rangeStart and rangeEnd:
+//
+//   - Start value: taken from the last point with T <= rangeStart,
+//     or, if none exists, the first point with T > rangeStart.
+//   - End value:   taken from the first point with T >= rangeEnd,
+//     or, if none exists, the last point with T < rangeEnd.
```

Comment:
These points are only added if they're within the lookback window, right?

eg. if `rangeStart` is T=10m, then the start value will be the first point before T=10m that has T > 5m, or else it'll be the first point after T=10m that has T < 15m